### PR TITLE
Actualize proposal after splitting immutable and persistent collections

### PR DESCRIPTION
**Pull request by ilya-g**

### CODE REVIEW

File: proposal.md [hunk 1]
```diff
@@ -1,136 +1,239 @@
...
+and implemented by the JDK collection implementations, such as `ArrayList`, `HashSet`, `HashMap` and other
+classes from `java.util` package.
+
+It is proposed to provide two families of interfaces that extend read-only collection interfaces:
+
+- Immutable collections: `ImmutableList`, `ImmutableSet`, etc. They specify by their contract the real immutability of their implementors.
+
+- Persistent collections: `PersistentList`, `PersistentSet`, etc. They extend immutable collections and provide efficient
+modification operations that return new instances of persistent collections with the modification applied. 
+The returned collections share parts of data structure with the original persistent collections.
```
Comment (qurbonzoda) [inline proposal.md]: This behaviour should not be forced as contract, it is an implementation detail. Small persistent lists actually do not share any parts.

Comment (qurbonzoda) [inline proposal.md]:
```suggestion
The returned collections may share parts of data structure with the original persistent collections.
```

File: proposal.md [hunk 2]
```diff
@@ -1,136 +1,239 @@
...
+
+For persistent collection interfaces there shall be provided the following implementations:
+
+- default persistent list
+- default persistent hash set preserving the insertion order
```
Comment (qurbonzoda) [inline proposal.md]: I am not sure if hash-based implementation detail should be exposed.
But, it's not a documentation, it's a proposal, so maybe it's ok.

Comment (ilya-g) [reply proposal.md]: Using object hash to store/retrieve object or key is an important aspect of the implementation. It limits what objects can be safely stored in these collections and retrieved back (those that have their hash unchanged after placing them in a map/set).

File: proposal.md (hunk 1 above)
Comment (ilya-g) [reply proposal.md]: Changed to "can share"

**Comment by qurbonzoda** [comment]
Are the added implementation details enough?

File: proposal.md [hunk 3]
```diff
@@ -199,18 +199,57 @@ implementing such methods.
...
 ### Persistent collection implementations
 
-TODO: describe some implementation details and performance characteristics of our persistent collection impls.
-
 #### Default persistent list
 
-#### Default persistent ordered hash set
+It's backed by a bit-mapped trie with branching factor of 32.
+`get` and `set` operations takes O(log<sub>32</sub>N) time, where N is the size of the instance the operations are apply on.
+To optimize frequently used _add last_ and _remove last_ operations rightmost leaf is referenced directly from the persistent list instance.
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply proposal.md]: Also time complexity of other list operations:
- insertion in the middle: single elements, multiple elements
- removing from the middle

[the remark this answers is not in the document]
Comment (qurbonzoda) [reply proposal.md]: done

File: proposal.md [hunk 4]
```diff
@@ -1,233 +1,372 @@
...
-  in a chain of operations.
+It's backed by a compressed hash-array mapped prefix-tree (a.k.a CHAMP). Every node has up to 32 children or entries.
+
+Time complexity of operations:
+- `get(key)`, `put(key, value)`, `remove(key)`, `remove(key, value)`, `containsKey(key)` - O(log<sub>32</sub>N) in average, but strongly depends on hash codes of the stored keys.
+- `containsValue(value)` - O(N).
+- `putAll(map)` - O(M log<sub>32</sub>N), where M is the number of elements added.
+- Iterating `keys`, `values`, `entries` - O(N).
+
+#### Default persistent ordered hash set
```
Comment (ilya-g) [inline proposal.md]: I believe it'd be better to describe ordered map first and then explain ordered set as a simplification of ordered map.

Comment (qurbonzoda) [reply proposal.md]: Done.

**Comment by ilya-g** [comment]
Merged manually to master

**Pull request closed without merging**